model/riscv_step.sail

```diff
@@ -37,36 +38,74 @@ union Step = {
   Step_Ext_Fetch_Failure  : ext_fetch_addr_error,
   Step_Fetch_Failure      : (virtaddr, ExceptionType),
   Step_Execute            : (ExecutionResult(Retire_Failure), instbits),
-  Step_Waiting            : unit,
+  Step_Waiting            : WaitReason,
 }
 
-function run_hart_waiting(step_no : int, exit_wait : bool, instbits : instbits) -> (Step, bool) = {
-  if shouldWakeForInterrupt() then {
-    if   get_config_print_instr()
-    then print_instr("interrupt exit from WAIT state at PC " ^ BitStr(PC));
+function run_hart_waiting(step_no : int, wr: WaitReason, instbits : instbits, exit_wait : bool) -> (Step, bool) = {
+  match (shouldWakeForInterrupt(), wr, valid_reservation(), exit_wait) {
+    // successfully interrupted waits
+    (true, _, _, _) => {
+      if  get_config_print_instr()
+      then print_instr("interrupt exit from WAIT state at PC " ^ BitStr(PC));
 
-    hart_state = HART_ACTIVE();
-    // The waiting instruction retires successfully.  The
-    // pending interrupts will be handled in the next step.
-    (Step_Execute(RETIRE_OK(), instbits), true)
-  } else if exit_wait then {
-    // There are no pending interrupts; transition out of the Wait
-    // as instructed.
-    if   get_config_print_instr()
-    then print_instr("forced exit from WAIT state at PC " ^ BitStr(PC));
+      // The waiting instruction retires successfully.  The
+      // pending interrupts will be handled in the next step.
+      hart_state = HART_ACTIVE();
+      (Step_Execute(RETIRE_OK(), instbits), true)
+    },
+    (_, WAIT_WRS_STO, false, _) => {
+      if   get_config_print_instr()
+      then print_instr("reservation invalid exit from WRS-STO state at PC " ^ BitStr(PC));
```

Comment:
Nit: It may be reservation set becoming invalid or an interrupt.

Reply:
See above.